# 1. Sposób na brute force'a

### Rys. pomocniczy
[![Rys. Pomocniczy](https://i.imgur.com/nm7hh6U.png)](https://imgur.com)

# Zalozenie: po scianie w gore pajak zawsze idzie pionowo

### Okreslmy zmienne

In [24]:
#spider variables
x_pajaka = 1
y_pajaka = 2

#fly variables
x_muchy = 2
y_muchy = 3

#room dimensions
x = 3   #scania na lewo to 0, sciana na prawo to 3
y = 4   #sciana tylna to 4, sciana przednia to 0
z = 2 #zakladamy dla prostoty ze pajak idzie po scianie zawsze pionowo bo to najkrotsza droga zeby pokonac wymiar wysokosci

### Tworzymy podzialke

In [25]:
#dzielimy podloge i sufit na n elementow, w tym wypadku powiedzmy ze zrobimy podzialke 0.1
import numpy as np
podzialka = 0.1
x_array = np.arange(0,x+podzialka,podzialka)
y_array = np.arange(0,y+podzialka,podzialka)
print ("X: \n",x_array, "\n")
print ("Y: \n",y_array, "\n")

X: 
 [0.  0.1 0.2 0.3 0.4 0.5 0.6 0.7 0.8 0.9 1.  1.1 1.2 1.3 1.4 1.5 1.6 1.7
 1.8 1.9 2.  2.1 2.2 2.3 2.4 2.5 2.6 2.7 2.8 2.9 3. ] 

Y: 
 [0.  0.1 0.2 0.3 0.4 0.5 0.6 0.7 0.8 0.9 1.  1.1 1.2 1.3 1.4 1.5 1.6 1.7
 1.8 1.9 2.  2.1 2.2 2.3 2.4 2.5 2.6 2.7 2.8 2.9 3.  3.1 3.2 3.3 3.4 3.5
 3.6 3.7 3.8 3.9 4. ] 



### Rozwazamy 4 scenariusze, pajak idzie na lewa, prawa, tylna lub przednia sciane. Iterujemy po kolei i liczymy odleglosci i wybieramy najmniejsza wartosc.

##### 1. Tylna sciana

In [26]:
#tylna sciana
import math

droga_calkowita_array1 = []  #tworzymy pusty array

for value in x_array:
    droga_podloga = math.sqrt((y-y_pajaka)**2+(value-x_pajaka)**2)
    droga_sciana = z
    droga_sufit = math.sqrt((y_muchy-y)**2+(x_muchy-value)**2)
    droga_calkowita = droga_podloga + droga_sciana + droga_sufit
    droga_calkowita_array1.append(droga_calkowita)

print(droga_calkowita_array1)

[6.47213595499958, 6.34026227530452, 6.2131919510512015, 6.091270302373312, 5.974857528193431, 5.8643284505408255, 5.76007285884564, 5.6624967883013415, 5.572025059405509, 5.48910531418193, 5.414213562373095, 5.34786084415745, 5.290599971710748, 5.243030403189039, 5.2057981844061745, 5.179586801558726, 5.165094263209011, 5.1629926609327645, 5.173869825572359, 5.19815878205822, 5.23606797749979, 5.287530004214754, 5.352184660656677, 5.429402739266368, 5.518344084573641, 5.618033988749895, 5.7274400739422, 5.845536511254707, 5.971349656901313, 6.103985249534116, 6.242640687119285]


In [27]:
#wybieramy minimalna wartosc
sciana_tylna = min(droga_calkowita_array1)
print("SCIANA TYLNA: \nMinimalna droga to:", round(sciana_tylna, 4), " czyli pierwiastek z: ", round(sciana_tylna**2,4))

SCIANA TYLNA: 
Minimalna droga to: 5.163  czyli pierwiastek z:  26.6565


##### 2. Przednia sciana

In [28]:
#przednia sciana
import math

droga_calkowita_array2 = []  #tworzymy pusty array

for value in x_array:
    droga_podloga = math.sqrt((0-y_pajaka)**2+(value-x_pajaka)**2)
    droga_sciana = z
    droga_sufit = math.sqrt((y_muchy-0)**2+(x_muchy-value)**2)
    droga_calkowita = droga_podloga + droga_sciana + droga_sufit
    droga_calkowita_array2.append(droga_calkowita)

print(droga_calkowita_array2)

[7.841619252963779, 7.744227400859072, 7.652637059760982, 7.567149939955043, 7.4880613017821105, 7.4156547790585154, 7.350196876886484, 7.291931386470032, 7.241074008504881, 7.197807501184171, 7.16227766016838, 7.134590392123244, 7.114810063476183, 7.102959201765541, 7.099019513592785, 7.10293407795794, 7.114610491866421, 7.133924696377976, 7.160725198528382, 7.194837423906858, 7.23606797749979, 7.284208646063392, 7.339040033612702, 7.400334774711579, 7.467860313231052, 7.541381265149109, 7.6206614031288105, 7.70546530983121, 7.795559748666747, 7.89071479749991, 7.99070478491457]


In [29]:
#wybieramy minimalna wartosc
sciana_przednia = min(droga_calkowita_array2)
print("SCIANA PRZEDNIA: \nMinimalna droga to:", round(sciana_przednia, 4), " czyli pierwiastek z: ", round(sciana_przednia**2,4))

SCIANA PRZEDNIA: 
Minimalna droga to: 7.099  czyli pierwiastek z:  50.3961


##### 3. Lewa sciana

In [30]:
#lewa sciana
import math

droga_calkowita_array3 = []  #tworzymy pusty array

for value in y_array:
    droga_podloga = math.sqrt((value-y_pajaka)**2+(0-x_pajaka)**2)
    droga_sciana = z
    droga_sufit = math.sqrt((y_muchy-value)**2+(x_muchy-0)**2)
    droga_calkowita = droga_podloga + droga_sciana + droga_sufit
    droga_calkowita_array3.append(droga_calkowita)

print(droga_calkowita_array3)

[7.841619252963779, 7.669874046120096, 7.50005613501445, 7.332367815613891, 7.1670401197826665, 7.004337756448419, 6.844564923771188, 6.688072077511307, 6.535263684645032, 6.38660687473185, 6.242640687119286, 6.103985249534115, 5.971349656901311, 5.845536511254707, 5.7274400739422, 5.618033988749895, 5.518344084573641, 5.429402739266368, 5.352184660656677, 5.287530004214755, 5.23606797749979, 5.19815878205822, 5.173869825572359, 5.162992660932764, 5.16509426320901, 5.179586801558726, 5.205798184406174, 5.243030403189039, 5.290599971710748, 5.34786084415745, 5.414213562373095, 5.489105314181929, 5.572025059405509, 5.6624967883013415, 5.760072858845639, 5.864328450540825, 5.974857528193431, 6.091270302373312, 6.213191951051202, 6.340262275304521, 6.47213595499958]


In [31]:
#wybieramy minimalna wartosc
sciana_lewa = min(droga_calkowita_array3)
print("SCIANA LEWA: \nMinimalna droga to:", round(sciana_lewa, 4), " czyli pierwiastek z: ", round(sciana_lewa**2,4))

SCIANA LEWA: 
Minimalna droga to: 5.163  czyli pierwiastek z:  26.6565


##### 4. Prawa sciana

In [32]:
#prawa sciana
import math

droga_calkowita_array4 = []  #tworzymy pusty array

for value in y_array:
    droga_podloga = math.sqrt((value-y_pajaka)**2+(x-x_pajaka)**2)
    droga_sciana = z
    droga_sufit = math.sqrt((y_muchy-value)**2+(x_muchy-x)**2)
    droga_calkowita = droga_podloga + droga_sciana + droga_sufit
    droga_calkowita_array4.append(droga_calkowita)

print(droga_calkowita_array4)

[7.99070478491457, 7.826195174862338, 7.663938558878444, 7.504116959458932, 7.346927350409963, 7.192582403567252, 7.041311123146741, 6.893359329172203, 6.748989952657034, 6.608483112025267, 6.47213595499958, 6.340262275304519, 6.2131919510512015, 6.091270302373312, 5.97485752819343, 5.8643284505408255, 5.76007285884564, 5.6624967883013415, 5.572025059405509, 5.48910531418193, 5.414213562373095, 5.34786084415745, 5.290599971710748, 5.243030403189039, 5.2057981844061745, 5.179586801558726, 5.165094263209011, 5.1629926609327645, 5.173869825572359, 5.19815878205822, 5.23606797749979, 5.287530004214754, 5.352184660656677, 5.429402739266368, 5.518344084573641, 5.618033988749895, 5.7274400739422, 5.845536511254707, 5.971349656901313, 6.103985249534116, 6.242640687119285]


In [33]:
#wybieramy minimalna wartosc
sciana_prawa = min(droga_calkowita_array4)
print("SCIANA PRAWA: \nMinimalna droga to:", round(sciana_prawa, 4), " czyli pierwiastek z: ", round(sciana_prawa**2,4))

SCIANA PRAWA: 
Minimalna droga to: 5.163  czyli pierwiastek z:  26.6565


### 5. Ktora sciana najlepsza i jaka najkrotsza droga? (przy zalozeniu ze pajak idzie w gore)

In [34]:
dlugosci = {
    "sciana_tylna": sciana_tylna,
    "sciana_lewa": sciana_lewa,
    "sciana_prawa": sciana_prawa,
    "sciana_przednia": sciana_przednia
}
odp = min(dlugosci.values())
sciana = [var for var, value in dlugosci.items() if value == odp][0]
print("Najkrotsza droga to: ",sciana, round(odp, 4) , " czyli pierwiastek z: ", round(odp**2,4))

Najkrotsza droga to:  sciana_lewa 5.163  czyli pierwiastek z:  26.6565


##### Mozna wybrac sciane lewa, prawa albo tylna, wszystkie maja taki sam dystans bo system wspolrzednych jest zamkniety.
##### Najwazniejsze zeby nie wybrac sciany tylnej bo pajak sie cofa i wydluza sobie droge

# Zalozenie ze pajak idzie na skos idac w gore (juz stwierdzilismy ze lewa, prawa, albo tylna sciana da taka sama minimalna droge, tylko zeby przedniej nie wybierac)

### Robimy podzialke dla podlogi, sciany i sufitu

In [35]:
#dzielimy podloge, sufit i sciane na n elementow, w tym wypadku powiedzmy ze zrobimy podzialke 0.1
import numpy as np
podzialka = 0.1
x_array = np.arange(0,x+podzialka,podzialka)
y_array = np.arange(0,y+podzialka,podzialka)
z_array = np.arange(0,z+podzialka, podzialka)
print ("X: \n",x_array, "\n")
print ("Y: \n",y_array, "\n")
print ("Z: \n",z_array, "\n")

X: 
 [0.  0.1 0.2 0.3 0.4 0.5 0.6 0.7 0.8 0.9 1.  1.1 1.2 1.3 1.4 1.5 1.6 1.7
 1.8 1.9 2.  2.1 2.2 2.3 2.4 2.5 2.6 2.7 2.8 2.9 3. ] 

Y: 
 [0.  0.1 0.2 0.3 0.4 0.5 0.6 0.7 0.8 0.9 1.  1.1 1.2 1.3 1.4 1.5 1.6 1.7
 1.8 1.9 2.  2.1 2.2 2.3 2.4 2.5 2.6 2.7 2.8 2.9 3.  3.1 3.2 3.3 3.4 3.5
 3.6 3.7 3.8 3.9 4. ] 

Z: 
 [0.  0.1 0.2 0.3 0.4 0.5 0.6 0.7 0.8 0.9 1.  1.1 1.2 1.3 1.4 1.5 1.6 1.7
 1.8 1.9 2. ] 



##### Iterujemy po krawedzi podlogi i dla kazdej iterowanej wartosci iterujemy po krawedzi sufita i liczymy odleglosc

In [38]:
#tylna sciana
import math

droga_calkowita_array = []  #tworzymy pusty array

for value in x_array:
    droga_podloga = math.sqrt((value-x_pajaka)**2 + (y-y_pajaka)**2)
    for value2 in x_array:

        droga_sciana = math.sqrt((value2-value)**2+(z-0)**2)
        droga_sufit = math.sqrt((x_muchy-value2)**2+(y_muchy-y)**2)
        droga_calkowita = droga_podloga + droga_sciana + droga_sufit
        droga_calkowita_array.append(droga_calkowita)


print(droga_calkowita_array)

[6.47213595499958, 6.3856574723082575, 6.305169129921368, 6.230751111447061, 6.1624720093482255, 6.100396428040615, 6.0445943326904255, 5.995151934227172, 5.952183835534922, 5.9158460721777715, 5.8863495173726745, 5.863972824309826, 5.849073582924479, 5.842095627448472, 5.84356947961559, 5.854101966249685, 5.87435063389983, 5.904979578072182, 5.94659668963309, 5.9996783844386234, 6.06449510224598, 6.141055539611879, 6.229085629682048, 6.328048759216479, 6.437200809289353, 6.555664084966109, 6.682502249840195, 6.816783062355448, 6.957622931803411, 7.104213372968869, 7.255832815336874, 6.431737636895999, 6.34026227530452, 6.254795687593609, 6.175454636501911, 6.10234228797312, 6.0355546631152395, 5.975189086163486, 5.921354468413914, 5.87418316516917, 5.833844017531784, 5.800556002265357, 5.774601602153291, 5.756338509535366, 5.746207539457622, 5.744733687290504, 5.752516331842767, 5.770204181373032, 5.798451565810326, 5.8378560723460255, 5.888883591472962, 5.951794064772875, 6.026585906

#### Wizualizacja iteracji
[![Rys. Pomocniczy](https://i.imgur.com/mcv5iIw.png)](https://imgur.com)

In [39]:
#wybieramy minimalna wartosc
droga_min = min(droga_calkowita_array)
print("Minimalna droga to:", round(droga_min, 4), " czyli pierwiastek z: ", round(droga_min**2,4))

Minimalna droga to: 5.099  czyli pierwiastek z:  26.0


### WNIOSEK: Widzimy ze pajak powinien isc po skosie zeby osiagnac najkrotsza droge która wynosi 5.099 j albo $\sqrt{26}$ j


# 2. Sposob matematyczny

Trzeba znaleźć funkcje matematyczną i znaleźć jej minimum za pomocą pochodnych

Wzór na droge calkowita:

*droga_calkowita = droga_podloga + droga_sciana + droga_sufit*
__________________________________________________________________________________________________

### 2.1 Wzor na podloge (idziemy na tylna sciane)

droga_podloga =  $\sqrt{(x_{i} - x_{pajaka})^2 +(y_{podlogi} - y_{pajaka})^2 }$

droga_podloga(x) = $\sqrt{(x - 1)^2 +(4 - 2)^2 }$ = $\sqrt{(x - 1)^2 + 4 }$

### 2.2 Wzor na sciane

droga_sciana =  $\sqrt{(x_{j} - x_{i})^2 +(z_{sufitu} - z_{podlogi})^2 }$

droga_sciana(x, y) = $\sqrt{(y - x)^2 +(2 - 0)^2 }$ = $\sqrt{(y - x)^2 + 4 }$

### 2.3 Wzor na sufit

droga_sufit =  $\sqrt{(x_{muchy} - x_{j})^2 +(y_{muchy} - y_{sufitu})^2 }$

droga_sufit(y) = $\sqrt{(2 - y)^2 +(3 - 4)^2 }$ = $\sqrt{(2 - y)^2 + 1 }$
_________________________________________________________________________________________

### <span style="color:green">2.4 Koncowy wzor:<span>

*f(x, y)* = $\sqrt{(x - 1)^2 + 4 }$ + $\sqrt{(y - x)^2 + 4 }$ + $\sqrt{(2 - y)^2 + 1 }$

#### <span style="color:green">2.4.1 Pochodne czastkowe<span>


f'<sub>x</sub> = $\frac{x-1}{\sqrt{x^2 - 2x +5}}$ + $\frac{x-y}{\sqrt{x^2 - 2yx + y^2 +4}}$

f'<sub>y</sub> = $\frac{y-2}{\sqrt{y^2 - 4y +5}}$ + $\frac{-x+y}{\sqrt{x^2 - 2yx + y^2 +4}}$


---------------------------------------------------------------------------------------------

##### 2.4.2 Przyrownujemy do zera i dodajemy do siebie:

$\frac{x-1}{\sqrt{x^2 - 2x +5}}$ + $\frac{x-y}{\sqrt{x^2 - 2yx + y^2 +4}}$ = 0

$\frac{y-2}{\sqrt{y^2 - 4y +5}}$ + $\frac{-x+y}{\sqrt{x^2 - 2yx + y^2 +4}}$ = 0

$\frac{x-1}{\sqrt{x^2 - 2x +5}}$ + $\frac{y-2}{\sqrt{y^2 - 4y +5}}$ = 0

---------------------------------------------------------------------------------------------

##### 2.4.3 Funkcje kwadratowe pod pierwiastkiem nie maja miejsc zerowych i sa zawsze dodatnie. Przerzucamy y na druga strone i podnosimy do kwadratu i pozbywamy sie wartosci bezwzglednej:


$\frac{x^2 -2x +1}{x^2-2x+5}$ = $\frac{y^2-4y+4}{y^2-4y+5}$

---------------------------------------------------------------------------------------------

##### 2.4.4 Mnożymy na skos i przerzucamy wszystko na lewa strone:

$x^2 - 2x - 15 - 4y^2 + 16y = 0$

---------------------------------------------------------------------------------------------

##### 2.4.5 Robimy rownanie po okregu:


$x^2 -2x +1 -16 -4(y^2 -4y+4)+16=0$

Przeksztalcamy:

$(x-1)^2-4(y-2)^2=0$

Jeszcze przeksztalcamy:

$(x-1)^2=4(y-2)^2$

---------------------------------------------------------------------------------------------

##### 2.4.6 Pierwiastkujemy:

$|x-1|=2|y-2|$

* **Dla y > 2**

$|x-1|=2(y-2)$

$y=\frac{|x-1|}{2}+2$

* **Dla y < 2**

$|x-1|=2(-y+2)$

$y=-\frac{|x-1|}{2}+2$

---------------------------------------------------------------------------------------------

* Dla $y > 2$ oraz $x > 1$

$x = 2y -3$

---------------------------------------------------------------------------------------------

* Dla $y > 2$ oraz $x < 1$

$x = -2y +5$

---------------------------------------------------------------------------------------------

* Dla $y < 2$ oraz $x > 1$

$x = -2y +5$

---------------------------------------------------------------------------------------------

* Dla $y < 2$ oraz $x < 1$

$x = -2y +3$

---------------------------------------------------------------------------------------------

### UWAGA1: x dotycza podlogi i sciany, y dotycza sciany i sufitu
### UWAGA2: nie mylic tutaj x za os x albo y za  os y, rownie dobrze mozna zamiast x, y napisac a, b, albo j, k

jezeli myslimy o tylnej scianie, iteracje sa tylko od 0 do 3, jesli prawa/lewa sciana to iteracje sa od 0 do 4

$x= <0,3>$

$y = <0,3>$ 

Spojrzmy na rys. pomocniczy i ocenmy ktore warunki nie maja sensu:

Jesli x <1 to nie ma to sensu, gdyz pajak bedzie sie cofal a nie zblizal do muchy kiedy idzie po podlodze

Jezeli zas y > 2 to tez nie ma sensu, bo mucha lezy na 2, po co ma isc dalej na ukos na scianie a potem sie cofac po suficie?

* Dla $y < 2$ oraz $x > 1$ to jedyny sluszny wybor, zatem:

$x = -2y +5$

---------------------------------------------------------------------------------------------

##### 2.4.7 Szukamy ekstremum, podstawiamy powyzsze do pochodnych czastkowych:

* $\frac{y-2}{\sqrt{y^2 - 4y +5}}$ + $\frac{2y -5+y}{\sqrt{(-2y +5)^2 - 2y(-2y +5) + y^2 +4}}$ = 0

lub


* $\frac{-2y +5-1}{\sqrt{(-2y +5)^2 - 2(-2y +5) +5}}$ + $\frac{-2y +5-y}{\sqrt{(-2y +5)^2 - 2y(-2y +5) + y^2 +4}}$ = 0

---------------------------------------------------------------------------------------------

Wybieram pierwsza opcje:

* $\frac{y-2}{\sqrt{y^2 - 4y +5}}$ + $\frac{3y -5}{\sqrt{4y^2-20y+25 + 4y^2 - 10y+ y^2 +4}}$ = 0

* $\frac{y-2}{\sqrt{y^2 - 4y +5}}$ + $\frac{3y -5}{\sqrt{9y^2-30y+29}}$ = 0

---------------------------------------------------------------------------------------------

Przerzucam stronami i podnosze do kwadratu:

* $\frac{y^2-4y+4}{|y^2 - 4y +5|}$   = $\frac{9y^2 -30y+25}{|9y^2-30y+29|}$

$|y^2 - 4y +5|$ jest zawsze dodatnie, $|9y^2-30y+29|$ jest rowniez dodatnie, brak miejsc zerowych, mozna pozbyc sie wartosci bezwzglednych

* $\frac{y^2-4y+4}{y^2 - 4y +5}$   = $\frac{9y^2 -30y+25}{9y^2-30y+29}$

---------------------------------------------------------------------------------------------

Mnozymy na krzyz i przenosimy na jedna strone:

$\cancel{9y^4}-\cancel{\textcolor{green}{30y^3}}+\textcolor{yellow}{29y^2}-\cancel{36y^3}+\textcolor{yellow}{120y^2}-\textcolor{orange}{116y}+\textcolor{yellow}{36y^2}-\textcolor{orange}{120y}+\textcolor{blue}{116} \textcolor{red}{=} \cancel{9y^4}-\cancel{\textcolor{green}{30y^3}}+\textcolor{yellow}{25y^2}-\cancel{36y^3}+\textcolor{yellow}{120y^2}-\textcolor{orange}{100y}+\textcolor{yellow}{45y^2}-\textcolor{orange}{150y}+\textcolor{blue}{125}$

$-5y^2+14y-9=0$

$(-y+1)*(5y-9) = 0$

---------------------------------------------------------------------------------------------

Punkty stacjonarne dla funkcji drogi to:
* $y=1$
* $x=3$

lub:

* $y=\frac{9}{5}=1.8$
* $x=\frac{7}{5}=1.4$

---------------------------------------------------------------------------------------------

##### 2.4.8 Podstawiamy do wzoru i liczymy droge: (nie liczymy drugich pochodnych, bo sa paskudne)

*f(x, y)* = $\sqrt{(x - 1)^2 + 4 }$ + $\sqrt{(y - x)^2 + 4 }$ + $\sqrt{(2 - y)^2 + 1 }$

*f(3, 1)* = $\sqrt{(3 - 1)^2 + 4 }$ + $\sqrt{(1 - 3)^2 + 4 }$ + $\sqrt{(2 - 1)^2 + 1 }$

*f(3, 1)* = $2$ + $2$ + $\sqrt{2}$

*f(3, 1)* = $4$ + $\sqrt{2}$ = $5.4142$ =$\sqrt{29.3}$ -> ***nie wiemy czy to max czy min***

---------------------------------------------------------------------------------------------

*$f(\frac{7}{5}, \frac{9}{5})$* = $\sqrt{(\frac{7}{5} - 1)^2 + 4 }$ + $\sqrt{(\frac{9}{5} - \frac{7}{5})^2 + 4 }$ + $\sqrt{(2 - \frac{9}{5})^2 + 1 }$

*$f(\frac{7}{5}, \frac{9}{5})$* = $\sqrt{\frac{4}{25} +4 }$ + $\sqrt{\frac{4}{25} + 4 }$ + $\sqrt{\frac{1}{25} + 1 }$

*$f(\frac{7}{5}, \frac{9}{5})$* = $\sqrt{\frac{104}{25} }$ + $\sqrt{\frac{104}{25} }$ + $\sqrt{\frac{26}{25} }$

*$f(\frac{7}{5}, \frac{9}{5})$* = $5.099$ = $\sqrt{26}$ -> ***jest mniejsze niz pierwszy wynik zatem to nasza odpowiedz (pokrywa sie z iteracja)***

---------------------------------------------------------------------------------------------